src/main/java/org/openstreetmap/atlas/geography/boundary/CountryBoundaryMap.java

```diff
@@ -104,91 +102,47 @@
-    // For backward compatibility
-    // TODO Remove once all files move to the new format
-    private static final String SPATIAL_INDEX_DELIMITER = "--";
-    private static final double MAX_AREA_FOR_NEAREST_NEIGHBOR = 100;
-    private static final double ANTIMERIDIAN = Longitude.MAX_VALUE;
+
     private static final int DEFAULT_MAXIMUM_POLYGONS_TO_SLICE_WITH = 2000;
     private static final int EXPANDED_MAXIMUM_POLYGONS_TO_SLICE_WITH = 25000;
     // Converters
-    private static final JtsMultiPolygonConverter JTS_MULTI_POLYGON_TO_POLYGON_CONVERTER = new JtsMultiPolygonConverter();
     private static final JtsPolyLineConverter JTS_POLYLINE_CONVERTER = new JtsPolyLineConverter();
     private static final JtsMultiPolygonToMultiPolygonConverter JTS_MULTI_POLYGON_TO_MULTI_POLYGON_CONVERTER = new JtsMultiPolygonToMultiPolygonConverter();
     private static final JtsPointConverter JTS_POINT_CONVERTER = new JtsPointConverter();
 
-    // WKT Helpers
-    private static final WKTWriter WKT_WRITER = new WKTWriter();
-
-    // The envelope of country boundary map.
-    private final Envelope envelope;
-
-    // The envelope of indexed area.
-    private Envelope gridIndexEnvelope;
-
     // Maps the ISO-3 country code to country boundary.
     private final MultiMap<String, Polygon> countryNameToBoundaryMap;
-
-    // The R-Tree containing all country boundaries for this CountryBoundaryMap.
-    private final STRtree rawIndex;
-
-    // We create a quad tree along the boundary of a country. Each leaf cell in the tree is
-    // guaranteed to only intersect a single country. This gives a performance boost during country
-    // code assignment for all features. To optimize the leaf cell lookup for areas or long ways, we
-    // store the quad tree cells into a R-Tree, which is the grid index.
-    private STRtree gridIndex;
+    private transient MultiMap<String, PreparedPolygon> countryNameToPreparedBoundaryPolyonMap;
 
     private boolean useExpandedPolygonLimit = true;
     private final transient GeometryPrecisionReducer reducer;
     private final CountryListTwoWayStringConverter countryListConverter = new CountryListTwoWayStringConverter();
 
     /**
-     * Collects leaf nodes of given {@link AbstractNode} into the given {@link MultiMap} from
-     * {@link Geometry} to {@link Envelope}s.
-     *
-     * @param node
-     *            Starting node for collection
-     * @param cells
-     *            {@link MultiMap} to save cells into
+     * @param countryGeometries
+     *            A list of {@link Geometry}s to check
+     * @return The set of country codes represented
      */
-    @SuppressWarnings("unchecked")
-    public static void collectCells(final AbstractNode node,
-            final MultiMap<Geometry, Envelope> cells)
+    public static Set<String> countryCodesIn(final Collection<?> countryGeometries)
     {
-        if (node.getLevel() > 0)
+        if (countryGeometries.iterator().next() instanceof Geometry)
```

Comment:
Instead of creating an iterator to check for object type you can directly say

```
return countryGeometries.stream().map(geometry -> {
    if(geometry instanceof Geometry)
        return getGeometryProperty((Geometry) geometry, ISOCountryTag.KEY));
    else if (geometry instanceof PreparedGeometry)
       return getGeometryProperty(((PreparedGeometry) geometry).getGeometry(), ISOCountryTag.KEY));
    return StringUtils.EMPTY;
}).filter(StringUtils::isNotEmpty).collect(Collectors.toSet());

Reply:
oh! this is great!! 👍